feat: add `formatting_func` support for TextClassification (#3599)
recognai/rubrix: docs/_source/guides/llms/practical_guides/fine_tune_others.md

3.  We then define an `ArgillaTrainer` instance with support for "openai", "setfit", "peft", "spacy" and "transformers".

> 

Reviewer: I guess this was an unintended line-break

> 

Author: ```suggestion
```

A brief shortcut that `RatingQuestion`s can be unified using a "majority"-, "min"-, "max"- or "disagreement"-strategy. Both `LabelQuestion`s and `MultiLabelQuestion`s can be resolved using a "majority"-, or "disagreement"-strategy.
````

#### Define a task

##### `text-label`-pair

> We offer the option to use default unification strategies and formatting based on a `text-label`-pair. Here we infer formatting information based on `*Field` and `*Question` names from the dataset. This is the easiest way to define a `TrainingTask` for text classification.

Reviewer: How does the asterisk work here before the `Field` and `Question`?

Author: ```suggestion
We offer the option to use default unification strategies and formatting based on a `text-label`-pair. Here we infer formatting information based on `TextField` and corresponding `Question` from the dataset. This is the easiest way to define a `TrainingTask` for text classification.
```

::::

##### `formatting_func`

We offer the option to provide a `formatting_func` to the `TrainingTask` or `TrainingTask`. This function is applied to each sample in the dataset and can be used for more advanced preprocessing and data formatting. The function should return a tuple of `(text, label)`.

```python
from argilla.feedback import FeedbackDataset, TrainingTask

dataset = FeedbackDataset.from_huggingface(
    repo_id="argilla/stackoverflow_feedback_demo"
)

def formatting_func(sample):
    text = sample["title"] + " " + sample["question"]
    values = [resp["value"] for resp in sample["title_question_fit"]]
    counter = Counter(values)
    if counter:
        most_common = counter.most_common()
        max_frequency = most_common[0][1]
        most_common_elements = [
            element for element, frequency in most_common if frequency == max_frequency
        ]
        label = random.choice(most_common_elements)
        return (text, label)
    else:
        return None

task = TrainingTask.for_text_classification(
    formatting_func=formatting_func,
)
```

#### Use ArgillaTrainer

> Next, we can use our `FeedbackDataset` and `TrainingTaskForTextClassification` to initialize our `argilla.ArgillaTrainer`. We support the frameworks "openai", "setfit", "peft", "spacy" and "transformers".

Reviewer: Is this meant to be `argilla.ArgillaTrainer` or `argilla.feedback.ArgillaTrainer`?

```suggestion
Next, we can use our `FeedbackDataset` and `TrainingTaskForTextClassification` to initialize our `argilla.feedback.ArgillaTrainer`. We support the frameworks "openai", "setfit", "peft", "spacy" and "transformers".
```

````{note}

> This is a newer version and can be imported via `from argilla.feedback import ArgillaTrainer`. The old trainer can be imported via `from argilla.training import ArgillaTrainer`. Our docs, contain some [additional information on usage of the ArgillaTrainer](/guides/train_a_model).

Reviewer: The path change should be reverted, right?

```suggestion
This is a newer version and can be imported via `from argilla.feedback import ArgillaTrainer`. The old trainer can be imported via `from argilla.training import ArgillaTrainer`. Our docs, contain some [additional information on usage of the ArgillaTrainer](../../train_a_model.html).
```

Author: No this is actually resolved correctly by sphinx internally.

trainer.update_config(num_train_epochs=2)
trainer.train(output_dir="my_awesone_model")
```
:::

:::{tab-item} formatting_func
```python
from argilla.feedback import ArgillaTrainer, FeedbackDataset, TrainingTask

dataset = FeedbackDataset.from_huggingface(
    repo_id="argilla/stackoverflow_feedback_demo"
)

def formatting_func(sample):
    text = sample["title"] + " " + sample["question"]
    values = [resp["value"] for resp in sample["title_question_fit"]]
    counter = Counter(values)
    if counter:
        most_common = counter.most_common()
        max_frequency = most_common[0][1]
        most_common_elements = [
            element for element, frequency in most_common if frequency == max_frequency
        ]
        label = random.choice(most_common_elements)
        return (text, label)
    else:

>         return None

Reviewer: How's the `return None` internally handled?

Author: within `_format_data` as with the `SFT`methods

Reviewer 2: ^^ `return None` is supported, useful if the sample was annotated as harmful or something.
